### PR TITLE
Add `AllTrialsCache` to `Study`

### PR DESCRIPTION
**Pull request by semiexp**
Based on [c-bata san's implementation](https://github.com/c-bata/optuna/commit/cdd508f4d4c40f7c2ae5ac8b99f9403aee69032e), we introduce `AllTrialsCache` to `Study` so that

- unnecessary query to DB can be avoided, and
- the same all-trials lists are used during one sampling. 

The differences between the original implementation and this are:

- The cache is populated lazily in this PR because not all samplers call `get_trials` (for example, `RandomSampler` do not).
- `use_cache=True` is used more widely.

### CODE REVIEW

**Comment by c-bata** [comment]
@amylase Could you review this PR?

**Review by amylase — APPROVED** [review]
LGTM. Added minor comments.

File: optuna/study/_optimize.py [hunk 1]
```diff
@@ -188,6 +188,7 @@ def _run_trial(
         optuna.storages.fail_stale_trials(study)
 
     trial = study.ask()
+    study._thread_local.cache_all_trials.activate(lambda: study.get_trials(deepcopy=False))
```
[the remark this answers is not in the document]
Comment (amylase) [reply optuna/study/_optimize.py:191]: As `study.get_trials` is a public & user facing method, we may want to hide `use_cache` argument as it is rather internal implementation details. How about adding a private method that behaves like `study.get_trials(use_cache=True)` and keep `study.get_trials` untouched?

**Comment by c-bata** [comment]
Let me close this PR since it was taken over at #4225. Thank you again for your contribution 💯

**Pull request closed without merging**